lib/functions/installs.bash

```diff
@@ -117,30 +117,189 @@ install_local_tool_versions() {
+    printf "%s\n" "$tools_installed"
+    return 0
+  fi
+  while IFS=' ' read -r tool_version; do
+    display_debug "_install_directory_tools '$search_path': found '$tool_version'"
+
+    # read one version from the file
+    IFS=' ' read -ra parts <<<"$tool_version"
+    local plugin_name
+    plugin_name=${parts[0]}
+
+    # skip if plugin is installed already
+    if [[ -n $(stringlist_contains "$tools_installed" "$plugin_name") ]]; then
+      display_debug "_install_directory_tools '$search_path': '$plugin_name' is already installed... skipping"
+      continue
+    fi
+
+    # install the versions
+    local plugin_version
+    for plugin_version in "${parts[@]:1}"; do
+      # install the version
+      display_none "$(install_tool_version "$plugin_name" "$plugin_version")"
+      tools_installed=$(printf "%s %s" "$tools_installed" "$plugin_name" | awk '{$1=$1};1')
+    done
+
+    # check if there is an environment override for it
+    # if so also install the environment version
+    local env_version
+    env_version=$(get_version_from_env "$plugin_name")
+    if [ -n "$env_version" ]; then
+      display_debug "_install_directory_tools '$search_path': $plugin_name: using environment override $env_version"
+      plugin_version=$env_version
+
+      # install the version
+      display_none "$(install_tool_version "$plugin_name" "$plugin_version")"
+      tools_installed=$(printf "%s %s" "$tools_installed" "$plugin_name" | awk '{$1=$1};1')
+    fi
+
+    display_debug "_install_directory_tools '$search_path': installed '$plugin_name':'$plugin_version' new state of tools_installed='$tools_installed'"
+  done <<<"$tool_versions"
+
+  printf "%s\n" "$tools_installed"
+}
+
+_install_directory_tools_legacy() {
+  local search_path=$1
+  local plugins_installed=$2
+  local tools_installed=$3
+
+  display_debug "_install_directory_tools_legacy '$search_path': resolving legacy files"
+  local plugin_name
+  for plugin_name in $plugins_installed; do
+    # skip if plugin is installed already
+    if [[ -n $(stringlist_contains "$tools_installed" "$plugin_name") ]]; then
+      display_debug "_install_directory_tools_legacy '$search_path': legacy_install $plugin_name: skipping as tool was already installed"
+      continue
+    fi
+
+    # extract plugin legacy information
+    local plugin_path
+    plugin_path=$(get_plugin_path "$plugin_name")
+    local legacy_list_filenames_script
+    legacy_list_filenames_script="${plugin_path}/bin/list-legacy-filenames"
+
+    # skip if no legacy_list_filenames_script available
+    if ! [[ -f "$legacy_list_filenames_script" ]]; then
+      display_debug "_install_directory_tools_legacy '$search_path': legacy_install $plugin_name: skipping as legacy files are not supported"
+      continue
+    fi
+
+    # extract plugin legacy filenames
+    local legacy_filenames=""
+    legacy_filenames=$("$legacy_list_filenames_script")
```

Comment:
For good measure quote here. I can't imagine any legacy file names would contain whitespace since these are typically filenames chosen for software developers, but it won't hurt.